packages/api/portfolios/environments/createEnvironment.ts

```diff
@@ -0,0 +1,109 @@
+import { errorHandlingMiddleware } from "../../utils/errorHandlingMiddleware";
+import createError from "http-errors";
+import { IpCheckerMiddleware } from "../../utils/ipLogging";
+import { validateRequestShape } from "../../utils/shapeValidator";
+
+/**
+ * Submits the environment of an application
+ *
+ * @param event - The POST request from API Gateway
+ */
+export async function baseHandler(
+  event: ApiGatewayEventParsed<IEnvironment>,
+  context?: Context
+): Promise<APIGatewayProxyResult> {
+  const setupResult = validateRequestShape<IEnvironment>(event);
+  const { portfolioId, applicationId } = setupResult.path;
+  const environmentBody = setupResult.bodyObject;
+
+  // set up database connection
+  const connection = await createConnection();
+  const insertedEnvironments: IEnvironmentCreate[] = [];
+
+  try {
+    // ensures both the portfolio and application exists
+    await connection.getRepository(Portfolio).findOneOrFail({
+      id: portfolioId,
+    });
+    const application = await connection.getRepository(Application).findOneOrFail({
+      id: applicationId,
+    });
+
+    // get all environment names for application
+    const environments = await connection
+      .getRepository(Environment)
+      .createQueryBuilder("environment")
+      .select(["environment.name"])
+      .where("environment.applicationId = :applicationId", { applicationId: application.id })
+      .getMany();
+
+    // ensure the environment name is unique for the application
+    // according to business rule 3.3
+    for (const environment of environments) {
+      if (environment.name === environmentBody.name) {
+        throw createError(400, "Duplicate Environment name in application", {
+          errorName: "DuplicateEnvironmentName",
+          environmentName: environmentBody.name,
+        });
+      }
+    }
+
+    const insertResult = await connection
+      .getCustomRepository(EnvironmentRepository)
+      .createEnvironment([{ ...environmentBody, application }]);
+
+    for (const environment of insertResult.identifiers) {
+      // the result from the insert method does not have the name property
+      // and an additional query after the insert is done to provide the correct shape
+      // to be returned to the client. This assumes that only one environment will be
+      // added at the moment but can be updated to handle more than one environment
+      const queryInsertedEnvironment = await connection.getCustomRepository(EnvironmentRepository).findOneOrFail({
+        select: [
+          "name",
+          "id",
+          "createdAt",
+          "updatedAt",
+          "archivedAt",
+          "administrators",
+          "contributors",
+          "readOnlyOperators",
+        ],
+        where: { id: environment.id },
+      });
+
+      insertedEnvironments.push(queryInsertedEnvironment);
+      console.log("Inserted Environment: " + JSON.stringify(queryInsertedEnvironment));
```

Comment:
Instead of inserting, and then doing a query, you could try to do a `.save` and then wrapping the response to only the fields you specified in your select

Reply:
Yeah I think I will try that in a separate refactor PR to keep the moment going.